File: commands/playlist.js
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../database/db');
const logger = require('../logger');

// Detect platform from URL
function detectPlatform(url) {
    if (url.includes('spotify.com')) return 'spotify';
    if (url.includes('youtube.com') || url.includes('youtu.be')) return 'youtube';
    if (url.includes('soundcloud.com')) return 'soundcloud';
    return 'other';
}

// Platform emojis
const PLATFORM_EMOJI = {
    spotify: '🟢',
    youtube: '🔴',
    soundcloud: '🟠',
    other: '🔗'
};

module.exports = {
    data: new SlashCommandBuilder()
        .setName('playlist')
        .setDescription('Save and manage your favorite playlists')
        .addSubcommand(sub =>
            sub.setName('save')
                .setDescription('Save a playlist for quick access')
                .addStringOption(opt =>
                    opt.setName('name')
                        .setDescription('A short name for this playlist (e.g., "chill", "workout")')
                        .setRequired(true)
                        .setMaxLength(50)
                )
                .addStringOption(opt =>
                    opt.setName('url')
                        .setDescription('The playlist URL (Spotify, YouTube, SoundCloud)')
                        .setRequired(true)
                )
        )
        .addSubcommand(sub =>
            sub.setName('play')
                .setDescription('Play one of your saved playlists')
                .addStringOption(opt =>
                    opt.setName('name')
                        .setDescription('Name of the saved playlist')
                        .setRequired(true)
                        .setAutocomplete(true)
                )
        )
        .addSubcommand(sub =>
            sub.setName('list')
                .setDescription('List all your saved playlists')
        )
        .addSubcommand(sub =>
            sub.setName('remove')
                .setDescription('Remove a saved playlist')
                .addStringOption(opt =>
                    opt.setName('name')
                        .setDescription('Name of the playlist to remove')
                        .setRequired(true)
                        .setAutocomplete(true)
                )
        ),

    // Handle autocomplete for playlist names
    async autocomplete(interaction) {
        const focusedValue = interaction.options.getFocused().toLowerCase();
        
        try {
            const [rows] = await db.pool.query(
                'SELECT name, platform FROM SavedPlaylists WHERE guildId = ? AND userId = ? ORDER BY name',
                [interaction.guildId, interaction.user.id]
            );
            
            const filtered = rows
                .filter(row => row.name.toLowerCase().includes(focusedValue))
                .slice(0, 25)
                .map(row => ({
                    name: `${PLATFORM_EMOJI[row.platform] || '🔗'} ${row.name}`,
                    value: row.name
                }));
            
            await interaction.respond(filtered);
        } catch (error) {
            logger.error('Playlist autocomplete error', { error: error.message });
            await interaction.respond([]);
        }
    },

    async execute(interaction, guildProfile) {
        const sub = interaction.options.getSubcommand();
        const client = interaction.client;

        // ========== SAVE ==========
        if (sub === 'save') {
            const name = interaction.options.getString('name').toLowerCase().trim();
            const url = interaction.options.getString('url').trim();
            
            // Validate URL format
            if (!url.startsWith('http://') && !url.startsWith('https://')) {
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.error} | Please provide a valid URL.`)
                    .setColor('#ff0000')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                return interaction.editReply({ embeds: [embed] });
            }
            
            // Validate name
            if (!/^[a-z0-9_-]+$/.test(name)) {
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.error} | Playlist name can only contain letters, numbers, dashes, and underscores.`)
                    .setColor('#ff0000')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                return interaction.editReply({ embeds: [embed] });
            }
            
            const platform = detectPlatform(url);
            
            try {
                // Check if user already has max playlists (limit: 25)
                const [countRows] = await db.pool.query(
                    'SELECT COUNT(*) as count FROM SavedPlaylists WHERE guildId = ? AND userId = ?',
                    [interaction.guildId, interaction.user.id]
                );
                
                if (countRows[0].count >= 25) {
                    const embed = new EmbedBuilder()
                        .setDescription(`${client.emotes.error} | You've reached the maximum of 25 saved playlists. Remove one first.`)
                        .setColor('#ff0000')
                        .setFooter({ text: 'The Pack', iconURL: client.logo });
                    return interaction.editReply({ embeds: [embed] });
                }
                
                // Insert or update
                await db.pool.query(
                    `INSERT INTO SavedPlaylists (guildId, userId, name, url, platform) 
                     VALUES (?, ?, ?, ?, ?)
                     ON DUPLICATE KEY UPDATE url = VALUES(url), platform = VALUES(platform)`,
                    [interaction.guildId, interaction.user.id, name, url, platform]
                );
                
                const embed = new EmbedBuilder()
                    .setTitle(`${client.emotes.success} | Playlist Saved!`)
                    .setDescription(`You can now play this playlist anytime with:\n\`/playlist play ${name}\``)
                    .addFields(
                        { name: 'Name', value: `\`${name}\``, inline: true },
                        { name: 'Platform', value: `${PLATFORM_EMOJI[platform]} ${platform}`, inline: true }
                    )
                    .setColor('#00ff00')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                
                return interaction.editReply({ embeds: [embed] });
                
            } catch (error) {
                logger.error('Failed to save playlist', { error: error.message });
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.error} | Failed to save playlist. Please try again.`)
                    .setColor('#ff0000')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                return interaction.editReply({ embeds: [embed] });
            }
        }

        // ========== PLAY ==========
        if (sub === 'play') {
            const name = interaction.options.getString('name').toLowerCase().trim();
            
            try {
                const [rows] = await db.pool.query(
                    'SELECT url, platform FROM SavedPlaylists WHERE guildId = ? AND userId = ? AND name = ?',
                    [interaction.guildId, interaction.user.id, name]
                );
                
                if (rows.length === 0) {
                    const embed = new EmbedBuilder()
                        .setDescription(`${client.emotes.error} | No playlist found with name \`${name}\`.\nUse \`/playlist list\` to see your saved playlists.`)
                        .setColor('#ff0000')
                        .setFooter({ text: 'The Pack', iconURL: client.logo });
                    return interaction.editReply({ embeds: [embed] });
                }
                
                const playlist = rows[0];
                
                // Get the play command and execute it with the URL
                const playCommand = client.commands.get('play');
                if (!playCommand) {
                    const embed = new EmbedBuilder()
                        .setDescription(`${client.emotes.error} | Play command not found.`)
                        .setColor('#ff0000')
                        .setFooter({ text: 'The Pack', iconURL: client.logo });
                    return interaction.editReply({ embeds: [embed] });
                }
                
                // Override the options to inject our saved URL
                const originalGetString = interaction.options.getString.bind(interaction.options);
                interaction.options.getString = (optionName) => {
                    if (optionName === 'song') return playlist.url;
                    return originalGetString(optionName);
                };
                
                // Execute play command with our URL
                return playCommand.execute(interaction, guildProfile);
                
            } catch (error) {
                logger.error('Failed to play saved playlist', { error: error.message });
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.error} | Failed to play playlist. Please try again.`)
                    .setColor('#ff0000')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                return interaction.editReply({ embeds: [embed] });
            }
        }

        // ========== LIST ==========
        if (sub === 'list') {
            try {
                const [rows] = await db.pool.query(
                    'SELECT name, url, platform, createdAt FROM SavedPlaylists WHERE guildId = ? AND userId = ? ORDER BY name',
                    [interaction.guildId, interaction.user.id]
                );
                
                if (rows.length === 0) {
                    const embed = new EmbedBuilder()
                        .setTitle('📋 Your Saved Playlists')
                        .setDescription('You don\'t have any saved playlists yet.\n\nSave one with:\n`/playlist save <name> <url>`')
                        .setColor('#ff006a')
                        .setFooter({ text: 'The Pack', iconURL: client.logo });
                    return interaction.editReply({ embeds: [embed] });
                }
                
                const playlistList = rows.map((row, index) => {
                    const emoji = PLATFORM_EMOJI[row.platform] || '🔗';
                    return `${emoji} **${row.name}** - [Link](${row.url})`;
                }).join('\n');
                
                const embed = new EmbedBuilder()
                    .setTitle(`📋 Your Saved Playlists (${rows.length}/25)`)
                    .setDescription(playlistList)
                    .setColor('#ff006a')
                    .setFooter({ text: 'The Pack • Use /playlist play <name> to play', iconURL: client.logo });
                
                return interaction.editReply({ embeds: [embed] });
                
            } catch (error) {
                logger.error('Failed to list playlists', { error: error.message });
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.error} | Failed to load playlists. Please try again.`)
                    .setColor('#ff0000')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                return interaction.editReply({ embeds: [embed] });
            }
        }

        // ========== REMOVE ==========
        if (sub === 'remove') {
            const name = interaction.options.getString('name').toLowerCase().trim();
            
            try {
                const [result] = await db.pool.query(
                    'DELETE FROM SavedPlaylists WHERE guildId = ? AND userId = ? AND name = ?',
                    [interaction.guildId, interaction.user.id, name]
                );
                
                if (result.affectedRows === 0) {
                    const embed = new EmbedBuilder()
                        .setDescription(`${client.emotes.error} | No playlist found with name \`${name}\`.`)
                        .setColor('#ff0000')
                        .setFooter({ text: 'The Pack', iconURL: client.logo });
                    return interaction.editReply({ embeds: [embed] });
                }
                
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.success} | Playlist \`${name}\` has been removed.`)
                    .setColor('#00ff00')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                
                return interaction.editReply({ embeds: [embed] });
                
            } catch (error) {
                logger.error('Failed to remove playlist', { error: error.message });
                const embed = new EmbedBuilder()
                    .setDescription(`${client.emotes.error} | Failed to remove playlist. Please try again.`)
                    .setColor('#ff0000')
                    .setFooter({ text: 'The Pack', iconURL: client.logo });
                return interaction.editReply({ embeds: [embed] });
            }
        }
    },
};

File: internal/commands/playlist.go
package commands

import (
	"context"
	"errors"
	"fmt"
	"regexp"
	"strings"

	"github.com/bwmarrin/discordgo"

	"github.com/OlliePCK/packbot/internal/storage"
	"github.com/OlliePCK/packbot/internal/style"
)

var playlistNameRe = regexp.MustCompile(`^[a-z0-9_-]+$`)

// detectPlatform mirrors the Node URL sniffing for the platform emoji.
func detectPlatform(url string) string {
	switch {
	case strings.Contains(url, "spotify.com"):
		return "spotify"
	case strings.Contains(url, "youtube.com"), strings.Contains(url, "youtu.be"):
		return "youtube"
	case strings.Contains(url, "soundcloud.com"):
		return "soundcloud"
	default:
		return "other"
	}
}

func platformEmoji(platform string) string {
	switch platform {
	case "spotify":
		return "🟢"
	case "youtube":
		return "🔴"
	case "soundcloud":
		return "🟠"
	default:
		return "🔗"
	}
}

// Playlist is /playlist — per-user saved playlists.
//
// NOTE: the `play` subcommand needs the music system; until the music batch
// lands it responds with a friendly not-yet message.
func Playlist(d Deps) *Command {
	return &Command{
		Def: &discordgo.ApplicationCommand{
			Name:        "playlist",
			Description: "Save and manage your favorite playlists",
			Options: []*discordgo.ApplicationCommandOption{
				{
					Type: discordgo.ApplicationCommandOptionSubCommand, Name: "save",
					Description: "Save a playlist for quick access",
					Options: []*discordgo.ApplicationCommandOption{
						{Type: discordgo.ApplicationCommandOptionString, Name: "name", Description: `A short name for this playlist (e.g., "chill", "workout")`, Required: true, MaxLength: 50},
						{Type: discordgo.ApplicationCommandOptionString, Name: "url", Description: "The playlist URL (Spotify, YouTube, SoundCloud)", Required: true},
					},
				},
				{
					Type: discordgo.ApplicationCommandOptionSubCommand, Name: "play",
					Description: "Play one of your saved playlists",
					Options: []*discordgo.ApplicationCommandOption{
						{Type: discordgo.ApplicationCommandOptionString, Name: "name", Description: "Name of the saved playlist", Required: true, Autocomplete: true},
					},
				},
				{Type: discordgo.ApplicationCommandOptionSubCommand, Name: "list", Description: "List all your saved playlists"},
				{
					Type: discordgo.ApplicationCommandOptionSubCommand, Name: "remove",
					Description: "Remove a saved playlist",
					Options: []*discordgo.ApplicationCommandOption{
						{Type: discordgo.ApplicationCommandOptionString, Name: "name", Description: "Name of the playlist to remove", Required: true, Autocomplete: true},
					},
				},
			},
		},
		Autocomplete: func(ctx context.Context, s *discordgo.Session, i *discordgo.InteractionCreate) ([]*discordgo.ApplicationCommandOptionChoice, error) {
			_, opts := subcommand(i)
			var focused string
			for _, o := range opts {
				if o.Focused {
					focused = strings.ToLower(o.StringValue())
				}
			}
			user := interactionUser(i)
			playlists, err := d.Store.ListPlaylists(ctx, i.GuildID, user.ID)
			if err != nil {
				return nil, err
			}
			var choices []*discordgo.ApplicationCommandOptionChoice
			for _, p := range playlists {
				if focused != "" && !strings.Contains(strings.ToLower(p.Name), focused) {
					continue
				}
				choices = append(choices, &discordgo.ApplicationCommandOptionChoice{
					Name:  platformEmoji(p.Platform) + " " + p.Name,
					Value: p.Name,
				})
				if len(choices) == 25 {
					break
				}
			}
			return choices, nil
		},
		Run: func(ctx context.Context, s *discordgo.Session, i *discordgo.InteractionCreate) error {
			sub, opts := subcommand(i)
			om := optionMap(opts)
			user := interactionUser(i)

			switch sub {
			case "save":
				name := strings.TrimSpace(strings.ToLower(om["name"].StringValue()))
				url := strings.TrimSpace(om["url"].StringValue())

				if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
					return Respond(s, i, style.ErrorEmbed("Please provide a valid URL."))
				}
				if !playlistNameRe.MatchString(name) {
					return Respond(s, i, style.ErrorEmbed("Playlist name can only contain letters, numbers, dashes, and underscores."))
				}

				platform := detectPlatform(url)
				err := d.Store.SavePlaylist(ctx, i.GuildID, user.ID, name, url, platform)
				if errors.Is(err, storage.ErrPlaylistLimit) {
					return Respond(s, i, style.ErrorEmbed(fmt.Sprintf("You've reached the maximum of %d saved playlists. Remove one first.", storage.MaxSavedPlaylists)))
				}
				if err != nil {
					return err
				}

				embed := &discordgo.MessageEmbed{
					Title:       style.Emotes.Success + " | Playlist Saved!",
					Description: fmt.Sprintf("You can now play this playlist anytime with:\n`/playlist play %s`", name),
					Color:       style.ColorSuccess,
					Footer:      style.Footer(),
					Fields: []*discordgo.MessageEmbedField{
						{Name: "Name", Value: "`" + name + "`", Inline: true},
						{Name: "Platform", Value: platformEmoji(platform) + " " + platform, Inline: true},
					},
				}
				return Respond(s, i, embed)

			case "play":
				name := strings.TrimSpace(strings.ToLower(om["name"].StringValue()))
				playlist, err := d.Store.GetPlaylist(ctx, i.GuildID, user.ID, name)
				if err != nil {
					return err
				}
				if playlist == nil {
					return Respond(s, i, style.ErrorEmbed(fmt.Sprintf("No playlist found with name `%s`.\nUse `/playlist list` to see your saved playlists.", name)))
				}
				if !requireMusic(d, s, i) {
					return nil
				}
				voiceChannel := requireVoice(s, i)
				if voiceChannel == "" {
					return nil
				}
				return playQuery(ctx, d, s, i, playlist.URL, voiceChannel, user)

			case "list":
				playlists, err := d.Store.ListPlaylists(ctx, i.GuildID, user.ID)
				if err != nil {
					return err
				}
				if len(playlists) == 0 {
					embed := &discordgo.MessageEmbed{
						Title:       "📋 Your Saved Playlists",
						Description: "You don't have any saved playlists yet.\n\nSave one with:\n`/playlist save <name> <url>`",
						Color:       style.ColorBrand,
						Footer:      style.Footer(),
					}
					return Respond(s, i, embed)
				}
				var lines []string
				for _, p := range playlists {
					lines = append(lines, fmt.Sprintf("%s **%s** - [Link](%s)", platformEmoji(p.Platform), p.Name, p.URL))
				}
				embed := &discordgo.MessageEmbed{
					Title:       fmt.Sprintf("📋 Your Saved Playlists (%d/%d)", len(playlists), storage.MaxSavedPlaylists),
					Description: strings.Join(lines, "\n"),
					Color:       style.ColorBrand,
					Footer:      &discordgo.MessageEmbedFooter{Text: "The Pack • Use /playlist play <name> to play", IconURL: style.LogoURL},
				}
				return Respond(s, i, embed)

			case "remove":
				name := strings.TrimSpace(strings.ToLower(om["name"].StringValue()))
				removed, err := d.Store.DeletePlaylist(ctx, i.GuildID, user.ID, name)
				if err != nil {
					return err
				}
				if !removed {
					return Respond(s, i, style.ErrorEmbed(fmt.Sprintf("No playlist found with name `%s`.", name)))
				}
				embed := &discordgo.MessageEmbed{
					Description: fmt.Sprintf("%s | Playlist `%s` has been removed.", style.Emotes.Success, name),
					Color:       style.ColorSuccess,
					Footer:      style.Footer(),
				}
				return Respond(s, i, embed)
			}
			return fmt.Errorf("unknown playlist subcommand %q", sub)
		},
	}
}
